### PR TITLE
Implement Rtree

### LINKED ISSUE #633: Prototype an R-tree for `SegmentPad`
An [R-tree](https://dl.acm.org/doi/10.1145/602259.602266) is a spatial index allowing fast search for geometry entities.  We will need the spatial indices for points, line segments, Bezier curves, and more geometry entities to be included in modmesh.

This issue is to track the first, prototype-ish implementation of R-tree for the line segments in `SegmentPad`.  The spatial index will be a foundation for high-level features.

The R-tree implementation should allow both 2 and 3 dimensional entities, but the tests can focus on 2D at the moment.

### PR DESCRIPTION
**Pull request by tigercosmos**
## Add R-tree Spatial Index Data Structure

This PR implements an R-tree spatial indexing data structure based on Guttman's seminal 1984 paper "R-TREES: A Dynamic Index Structure for Spatial Searching". The implementation provides efficient spatial queries for both 2D and 3D geometric objects.

### What is this PR?

This PR adds a generic, header-only R-tree implementation with the following components:

- **`BoundingBox2D` and `BoundingBox3D`**: Axis-aligned bounding box structures for 2D and 3D spaces
- **`RTree` template class**: A flexible R-tree implementation supporting custom value types and configurable node capacity
- **Comprehensive test suite**: 20+ test cases covering various scenarios and edge cases

The implementation follows the original R-tree algorithm including:
- Dynamic insertion with node splitting
- Spatial search with bounding box queries
- Item deletion with tree condensation
- Automatic tree height adjustment

#### Basic Setup

Define your data type and implement a value operations struct:

```cpp
struct Point2D {
    double x, y;
    bool operator==(Point2D const& other) const {
        return x == other.x && y == other.y;
    }
};

struct Point2DValueOps {
    static modmesh::BoundingBox2D get_bounding_box(Point2D const& item) {
        return modmesh::BoundingBox2D(item.x, item.y, item.x, item.y);
    }
    
    static modmesh::BoundingBox2D get_group_bounding_box(std::vector<Point2D> const& items) {
        // Calculate bounding box for all items
        // ...
    }
};
```

#### Creating and Using an R-tree

```cpp
// Create R-tree with custom max items per node (default is 64)
RTree<Point2D, BoundingBox2D, Point2DValueOps, 16> rtree;

// Insert items
rtree.insert(Point2D{1.0, 1.0});
rtree.insert(Point2D{2.0, 2.0});
rtree.insert(Point2D{3.0, 3.0});

// Search for items in a region
std::vector<Point2D> results;
rtree.search(BoundingBox2D(1.5, 1.5, 2.5, 2.5), results);

// Remove items
rtree.remove(Point2D{2.0, 2.0});
```

#### 3D Support

The same API works for 3D objects:

```cpp
RTree<Point3D, BoundingBox3D, Point3DValueOps> rtree_3d;
rtree_3d.insert(Point3D{1.0, 1.0, 1.0});
rtree_3d.search(BoundingBox3D(0.0, 0.0, 0.0, 5.0, 5.0, 5.0), results);
```

### Test Coverage

The test suite (`test_nopython_rtree.cpp`) provides comprehensive coverage with 20 test cases:

**Basic Operations:**
- `basic_operations` - Insert, search, and remove operations in 2D
- `basic_operations_3d` - Basic operations for 3D points
- `empty_tree_operations` - Handling empty trees
- `single_point_operations` - Single item edge cases

**Complex Scenarios:**
- `complex_operations` - Multiple insertions with forced node splits
- `large_scale_insertion_2d/3d` - Stress testing with 500-1000 items
- `random_insertion_and_search_2d/3d` - Random data with validation
- `stress_test_insert_remove_2d` - Heavy insertion/deletion cycles

**Edge Cases:**
- `overlapping_search_regions` - Overlapping query regions
- `boundary_conditions` - Exact boundary matching
- `scattered_points_2d` - Widely distributed data
- `duplicate_points_2d` - Handling duplicate values
- `non_overlapping_searches_2d` - Disjoint query regions

**Correctness Validation:**
- `multiple_removals_2d` - Sequential removal operations
- `sequential_removal_3d` - Tree structure after deletions
- `corner_cases_3d` - Special 3D configurations

All tests validate correctness by comparing search results against brute-force expected counts and ensuring tree integrity after insertions and deletions.

### Files Changed

- `cpp/modmesh/universe/rtree.hpp` - R-tree implementation (new)
- `gtests/test_nopython_rtree.cpp` - Comprehensive test suite (new)

## AI Usage

Using Copilot + Claude 4.5
1. Create test cases
2. Refine the code during the development
3. Generate the PR description

### CODE REVIEW

File: cpp/modmesh/universe/rtree.hpp [hunk 1]
```diff
@@ -0,0 +1,700 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// This library implement R-tree based on "R-TREES" paper by Guttman in 1984
```
Comment (tigercosmos) [inline cpp/modmesh/universe/rtree.hpp]: The implementation is based on the paper's algorithm. Functions name also follow the names in the paper.

File: cpp/modmesh/universe/rtree.hpp [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+#include <limits>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+namespace modmesh
+{
+
+/// Bounding box for 2D objects, e.g., Point2D, Segment2D, Triangle2D, etc.
+struct BoundingBox2D
```
Comment (tigercosmos) [inline cpp/modmesh/universe/rtree.hpp]: Only `BoundingBox2D` and `BoundingBox3D` now. Cannot support N-dim yet.

File: cpp/modmesh/universe/rtree.hpp [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+    }
+};
+
+/// R-tree implementation for spatial index based on Guttman's R-tree paper
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated with T
+/// @tparam ValueOps Value operations traits for T and BoundingBoxType
+/// @tparam MAX_ITEMS_PER_NODE Maximum number of items per R-tree node
+template <typename T, typename BoundingBoxType, typename ValueOps = RTreeValueOps<T, BoundingBoxType>, int MAX_ITEMS_PER_NODE = 64>
+class RTree
```
Comment (tigercosmos) [inline cpp/modmesh/universe/rtree.hpp:291]: Main implementation.

File: gtests/test_nopython_rtree.cpp [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+            min_z = std::min(min_z, item.z);
+            max_x = std::max(max_x, item.x);
+            max_y = std::max(max_y, item.y);
+            max_z = std::max(max_z, item.z);
+        }
+        return modmesh::BoundingBox3D(min_x, min_y, min_z, max_x, max_y, max_z);
+    }
+};
+
+TEST(RTree, basic_operations)
```
Comment (tigercosmos) [inline gtests/test_nopython_rtree.cpp:118]: Check the test cases to see how to use.

File: gtests/test_nopython_rtree.cpp [hunk 5]
```diff
@@ -0,0 +1,475 @@
+#include <gtest/gtest.h>
+#include <modmesh/universe/rtree.hpp>
+
+#include <random>
+#ifdef Py_PYTHON_H
+#error "Python.h should not be included."
+#endif
+
+// Define a simple value operations for 2D points
+struct Point2D
```
Comment (tigercosmos) [inline gtests/test_nopython_rtree.cpp:38]: Use simple class to test first. We can add modmesh classes in the tests later.

**Comment by tigercosmos** [comment]
@yungyuc  the PR is ready for review.

File: cpp/modmesh/universe/rtree.hpp (hunk 1 above)
Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp]: Add DOI link to the reference.

File: cpp/modmesh/universe/rtree.hpp (hunk 2 above)
Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp]: Rename `BoundingBox*` to `BoundBox*`.  It will be frequently used and a shorter name should be used.  Use `*2d` and `*3d` (lower case) like those for `Point*`, `Segment*`, `Triangle*`, and `Bezier*`.

Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp]: Make it a class template deriving from `NumberBase` and take a template argument for value type.  We will need different types of real number other than just double.

File: cpp/modmesh/universe/rtree.hpp [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+#include <type_traits>
+#include <vector>
+
+namespace modmesh
+{
+
+/// Bounding box for 2D objects, e.g., Point2D, Segment2D, Triangle2D, etc.
+struct BoundingBox2D
+{
+    double min_x;
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Keep `BoundBox` member data private.  We may not add more member data, but the order definition may change.

Provide accessors.

File: cpp/modmesh/universe/rtree.hpp [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+    BoundingBox2D(double min_x_in, double min_y_in, double max_x_in, double max_y_in)
+        : min_x(min_x_in)
+        , min_y(min_y_in)
+        , max_x(max_x_in)
+        , max_y(max_y_in)
+    {
+    }
+
+    /// check if this bounding box intersects another bounding box
+    bool intersects(BoundingBox2D const & other) const
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Do not name it `intersects` because it is ambiguous.  Use `overlap`.  Do not use singular form of the verb.  Use base form.

File: cpp/modmesh/universe/rtree.hpp [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+    {
+        return !(other.min_x > max_x || other.max_x < min_x || other.min_y > max_y || other.max_y < min_y);
+    }
+
+    /// check if this bounding box contains another bounding box
+    bool contains(BoundingBox2D const & other) const
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Use base form.

File: cpp/modmesh/universe/rtree.hpp [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+    {
+        return (other.min_x >= min_x && other.max_x <= max_x && other.min_y >= min_y && other.max_y <= max_y);
+    }
+
+    double get_area() const
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Rename to `calc_area()` since it is not a getter.

File: cpp/modmesh/universe/rtree.hpp [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+    {
+        return (max_x - min_x) * (max_y - min_y);
+    }
+
+    /// calculate the area enlargement needed to include the addition bounding box
+    double get_enlargement(BoundingBox2D const & addition) const
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: This API is too intrusive and should not be added.  Just use your static function `merge()` at the consuming sites:

```cpp
double diff_area = BoundBox::merge(box, obox).calc_area() - box.calc_area();
```

File: cpp/modmesh/universe/rtree.hpp [hunk 11]
```diff
@@ -0,0 +1,700 @@
...
+    /// expand the bounding box to include another bounding box
+    void expand(BoundingBox2D const & other)
+    {
+        min_x = std::min(min_x, other.min_x);
+        min_y = std::min(min_y, other.min_y);
+        max_x = std::max(max_x, other.max_x);
+        max_y = std::max(max_y, other.max_y);
+    }
+
+    /// merge two bounding boxes into a new bounding box
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Provide some usage example in the comments.

File: cpp/modmesh/universe/rtree.hpp [hunk 12]
```diff
@@ -0,0 +1,700 @@
...
+
+    using RTreeNodeType = RTreeNode<T, BoundingBoxType, ValueOpsType>;
+    std::vector<std::unique_ptr<RTreeNodeType>> nodes;
+
+    RTreeNode(BoundingBoxType const & bbox_in)
+        : bbox(bbox_in)
+    {
+    }
+
+    ~RTreeNode() = default;
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: I have a habit to add destructor after constructor.  FYI.  I don't mind you use a different convention.

File: cpp/modmesh/universe/rtree.hpp [hunk 13]
```diff
@@ -0,0 +1,700 @@
...
+    RTreeNode(RTreeNode const &) = delete;
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp:222]: Delete the default constructor.  I think it's automatically deleted but cannot be sure.  Don't make a maintainer guess.

File: cpp/modmesh/universe/rtree.hpp [hunk 14]
```diff
@@ -0,0 +1,700 @@
...
+    RTreeNode & operator=(RTreeNode const &) = delete;
+    RTreeNode(RTreeNode &&) = delete;
+    RTreeNode & operator=(RTreeNode &&) = delete;
+
+    template <bool IS_LEAF>
+    auto & get_children()
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp:229]: I am wary about the auto return type, but we can keep it here for a while before it starts to bite.

File: cpp/modmesh/universe/rtree.hpp [hunk 15]
```diff
@@ -0,0 +1,700 @@
...
+            return items;
+        }
+        else
+        {
+            return nodes;
+        }
+    }
+
+    /// Recalculate bounding box based on contained items and nodes
+    void recalculate_bounding_box()
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Rename everything about bounding box to bound box for consistency.

File: cpp/modmesh/universe/rtree.hpp [hunk 16]
```diff
@@ -0,0 +1,700 @@
...
+        bbox = result_box;
+    }
+}; /* end of struct RTreeNode */
+
+/// R-tree implementation for spatial index based on Guttman's R-tree paper
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated with T
+/// @tparam ValueOps Value operations traits for T and BoundingBoxType
+/// @tparam MAX_ITEMS_PER_NODE Maximum number of items per R-tree node
+template <typename T, typename BoundingBoxType, typename ValueOps = RTreeValueOps<T, BoundingBoxType>, int MAX_ITEMS_PER_NODE = 64>
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Rename `BoundingBoxType` as `B`.

Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp]: The real number value type should be named as `T`.  So the element type in the R-tree should be name as `E`.

File: cpp/modmesh/universe/rtree.hpp [hunk 17]
```diff
@@ -0,0 +1,700 @@
...
+    {
+    }
+    ~RTree() = default;
+    RTree(RTree const &) = delete;
+    RTree & operator=(RTree const &) = delete;
+    RTree(RTree &&) = delete;
+    RTree & operator=(RTree &&) = delete;
+
+    /// Insert item into R-tree
+    void insert(T const & item)
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: It's OK to still name the argument `item` after changing the type argument as `E`.  E is for element, but it is not wrong to refer to item.

Item is a shorter word than element.  I do not dislike it.

File: cpp/modmesh/universe/rtree.hpp [hunk 18]
```diff
@@ -0,0 +1,700 @@
...
+
+/// R-tree node structure
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated
+/// @tparam ValueOps Value operations traits for T and BoundingBoxType
+template <typename T, typename BoundingBoxType, typename ValueOpsType>
+struct RTreeNode
+{
+    BoundingBoxType bbox;
+    std::vector<T> items;
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: We may need better memory management and optimization for pointer operations than `std::vector`.  Keep it in mind and improve in the future.

File: cpp/modmesh/universe/rtree.hpp [hunk 19]
```diff
@@ -0,0 +1,700 @@
...
+}; /* end of struct RTreeValueOps */
+
+/// R-tree node structure
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated
+/// @tparam ValueOps Value operations traits for T and BoundingBoxType
+template <typename T, typename BoundingBoxType, typename ValueOpsType>
+struct RTreeNode
+{
+    BoundingBoxType bbox;
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Yep.  `bbox` is how we name things like this.

File: gtests/test_nopython_rtree.cpp (hunk 5 above)
Comment (yungyuc) [reply gtests/test_nopython_rtree.cpp:38]: This is good enough.  It will not be likely to significantly enrich this file.

Most testing will be done in Python.  You only need to cover very low-level API that will not be exposed to Python in the C++ test.

It is time-consuming to create test cases for geometry in C++.  You will write a lot of Python.

File: gtests/test_nopython_rtree.cpp [hunk 20]
```diff
@@ -0,0 +1,475 @@
+#include <gtest/gtest.h>
```
Comment (yungyuc) [inline gtests/test_nopython_rtree.cpp:29]: Copyright header and vim modeline footer.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Pretty nice.

General notes:

- Only name a member function `get`/`set` if it is a getter or setter.  When it calculates, prefix the name with `calc_`.
- While bounding box is the correct way to call a bounding box, for brevity, we should use bound box in code, which is not horribly wrong English.  Use bound box consistently.
- It's good to focus on C++ in this PR.

Points to address:

- [x] Add DOI link to reference.
- [x] Rename `BoundingBox*` to `BoundBox*` and use `*2d` and `*3d` (lower case).
- [x] Make `BoundBox*` a class template deriving accepting different types of real number than just double.
- [x] Make `BoundBox*` member data private and provide accessors.
- [x] Rename `intersects()` as `overlap()`.  Do base form instead of singular form of the verb.
- [x] Rename `get_area()` as `calc_area()`.
- [x] Remove `get_enlargement()`.
- [x] Explicitly delete the default constructor of `RTreeNode`.  Don't make a maintainer guess.
- [x] Rename the template argument `BoundingBoxType` to `RTree` as `B`.  The real number value type should be named as `T`.  The R-tree element (item) type should be named as `E`.
- [x] Copyright header and vim modeline footer for `test_nopython_rtree.cpp`.

This PR is for issue #633 .

**Comment by tigercosmos** [comment]
@yungyuc I fixed all issues. Please take a look.

File: cpp/modmesh/universe/rtree.hpp [hunk 21]
```diff
@@ -0,0 +1,709 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// This library implement R-tree based on "R Trees: A Dynamic Index Structure for Spatial Searchin" by Guttman in 1984
+// DOI:10.1145/971697.602266
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Use DOI link as I [mentioned](https://github.com/solvcon/modmesh/pull/637#discussion_r2552343480): https://doi.org/10.1145/971697.602266

File: cpp/modmesh/universe/rtree.hpp [hunk 22]
```diff
@@ -0,0 +1,709 @@
...
+        m_min_y = std::min(m_min_y, other.m_min_y);
+        m_max_x = std::max(m_max_x, other.m_max_x);
+        m_max_y = std::max(m_max_y, other.m_max_y);
+    }
+
+    /// merge two bounding boxes into a new bounding box
+    /// when merging, the resulting bounding box will be the minimum bounding box that contains both a and b
+    /// for example, if a is (1,1,3,3) and b is (2,2,4,4), the merged bounding box will be (1,1,4,4)
+    /// @return the merged bounding box of a and b
+    static BoundBox2d merge(BoundBox2d const & a, BoundBox2d const & b)
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: You don't need the static `merge()` because `expand()` can do the same thing:

```cpp
auto c = BoundBox2d(a, b);
auto c = BoundBox2d(a).expand(b);
```

Remove `merge()` at the moment.  If this pattern occurs frequently, we can add `merge()` in the future.

File: cpp/modmesh/universe/rtree.hpp [hunk 23]
```diff
@@ -0,0 +1,709 @@
...
+#include <type_traits>
+#include <vector>
+
+namespace modmesh
+{
+
+/// Bounding box for 2D objects, e.g., Point2d, Segment2d, Triangle2d, etc.
+/// @tparam T floating-point type
+template <typename T>
+struct BoundBox2d : public NumberBase<int32_t, T>
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: Use `class` instead of `struct` to make it obvious that the member data should be private.

File: cpp/modmesh/universe/rtree.hpp [hunk 24]
```diff
@@ -0,0 +1,709 @@
...
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+namespace modmesh
+{
+
+/// Bounding box for 2D objects, e.g., Point2d, Segment2d, Triangle2d, etc.
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp]: We do not have `Point2d`, `Segment2d`, `Triangle2d`.  We have `*3d` of the third axis unused.  Update the comment.

File: cpp/modmesh/universe/rtree.hpp [hunk 25]
```diff
@@ -0,0 +1,709 @@
...
+        return (other.m_min_x >= m_min_x && other.m_max_x <= m_max_x && other.m_min_y >= m_min_y && other.m_max_y <= m_max_y);
+    }
+
+    value_type calc_area() const
+    {
+        return (m_max_x - m_min_x) * (m_max_y - m_min_y);
+    }
+
+    /// expand the bounding box to include another bounding box
+    void expand(BoundBox2d const & other)
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp:103]: You can consider to add:

```cpp
Boundbox2d expand(BoundBox2d const & other) const
{
    return BoundBox2d(this).expand(other);
}
```

If it works, it will make the following two lines equivalent:

```cpp
auto c = BoundBox2d(a, b);
auto c = a.expand(b);  // calls the const version expand().
```

But I can't remember clearly if this kind of overloading works.  And a downside is the ambiguity by overloading (if it works).

File: cpp/modmesh/universe/rtree.hpp [hunk 26]
```diff
@@ -0,0 +1,709 @@
...
+{
+
+/// Bounding box for 2D objects, e.g., Point2d, Segment2d, Triangle2d, etc.
+/// @tparam T floating-point type
+template <typename T>
+struct BoundBox2d : public NumberBase<int32_t, T>
+{
+    using value_type = T;
+
+private:
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp:54]: Add same-name accessors.  Each value needs one.  You can consider to add const accessors first and leave non-const in the future.

File: cpp/modmesh/universe/rtree.hpp [hunk 27]
```diff
@@ -0,0 +1,709 @@
...
+            return items;
+        }
+        else
+        {
+            return nodes;
+        }
+    }
+
+    /// Recalculate bounding box based on contained items and nodes
+    void recalculate_bound_box()
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp:242]: `recalc_` instead of the long name?  The long name is OK though.

**Review by yungyuc — CHANGES_REQUESTED** [review]
An important guideline of designing fundamental library API like the bounding box and r-tree is to start with minimum.  Many "conveniences" will make the code base unmaintainable.

As we make applications on top of the fundamental geometry library, new API will emerge quickly.  We should use the most compact set to start.

- [x] Remove `BoundBox2d::merge()` static function.
- [x] Fix the comments on `Point2d`, `Segment2d`, `Triangle2d` (we do not have them).  Say that we have `*3d` with the third axis unused.
- [x] Missing accessors.

File: cpp/modmesh/universe/rtree.hpp (hunk 21 above)
Comment (tigercosmos) [reply cpp/modmesh/universe/rtree.hpp]: I see. Sorry, misunderstood before.

File: cpp/modmesh/universe/rtree.hpp (hunk 22 above)
Comment (tigercosmos) [reply cpp/modmesh/universe/rtree.hpp]: Sure. The only drawback is that we need to use two lines for a new bound box (copy first, then expand).

File: cpp/modmesh/universe/rtree.hpp (hunk 25 above)
Comment (tigercosmos) [reply cpp/modmesh/universe/rtree.hpp:103]: I added the new constructor `BoundBox2d(a, b);` .

I would like to keep the current `expand` function mutable because it is intended to change the self's bounding box. That's why there was a `merge` static function before.

File: cpp/modmesh/universe/rtree.hpp (hunk 26 above)
Comment (tigercosmos) [reply cpp/modmesh/universe/rtree.hpp:54]: Not sure if we even need the accessors, but I can add them.

File: cpp/modmesh/universe/rtree.hpp (hunk 27 above)
Comment (tigercosmos) [reply cpp/modmesh/universe/rtree.hpp:242]: I intended to not use abbreviation if the naming is not in convention or industry standard. Most of the time, we can find the code easily with the full name.

File: cpp/modmesh/universe/rtree.hpp (hunk 24 above)
Comment (tigercosmos) [reply cpp/modmesh/universe/rtree.hpp]: I meant we can have use 2D objects if they exist. But I agree maybe it will confuse people.

File: cpp/modmesh/universe/rtree.hpp (hunk 22 above)
Comment (tigercosmos) [reply cpp/modmesh/universe/rtree.hpp]: I like the idea to have a constructor to take two bbox as arguments. Use the new ctor and remve `merge`.

**Comment by tigercosmos** [comment]
@yungyuc  Please check [8d6cd69](https://github.com/solvcon/modmesh/pull/637/commits/8d6cd69a4d27c57a5bbbddc55a1903b6ebff5038) for the fixes.

File: cpp/modmesh/universe/rtree.hpp (hunk 26 above)
Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp:54]: Next time, if you do not want to add them, leave a comment saying that.

I am sure the accessors will be used.  Python wrapper is one case.  But the timing does not need to be now.

File: cpp/modmesh/universe/rtree.hpp (hunk 21 above)
Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp]: Not a problem.

File: cpp/modmesh/universe/rtree.hpp (hunk 22 above)
Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp]: It works too.

File: cpp/modmesh/universe/rtree.hpp [hunk 28]
```diff
@@ -0,0 +1,722 @@
...
+    BoundBox2d(value_type min_x_in, value_type min_y_in, value_type max_x_in, value_type max_y_in)
+        : m_min_x(min_x_in)
+        , m_min_y(min_y_in)
+        , m_max_x(max_x_in)
+        , m_max_y(max_y_in)
+    {
+    }
+
+    /// construct a bounding box that encloses two bounding boxes
+    BoundBox2d(const BoundBox2d & a, const BoundBox2d & b)
```
Comment (yungyuc) [inline cpp/modmesh/universe/rtree.hpp:72]: This is ambiguous, but ok for now.  If we need different behaviors in the future, new arguments can be added.

File: cpp/modmesh/universe/rtree.hpp (hunk 24 above)
Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp]: Right, the non-existing class names confuse people.  Saying 2D objects suffices.

File: cpp/modmesh/universe/rtree.hpp (hunk 25 above)
Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp:103]: I think the const and non-const overload can co-exist, but using a constructor to merge is OK too.  Having a single non-const `expand()` is clearer.  I like it.

File: cpp/modmesh/universe/rtree.hpp (hunk 27 above)
Comment (yungyuc) [reply cpp/modmesh/universe/rtree.hpp:242]: `calc_` and `recalc_` is my convention, and it's actually a convention in certain industries.  Using the two is OK.

I don't mind it is still named as `recalculate_` for now.  We will change in the future.  When starting to code in Python, soon you will realize how cumbersome the long name is.

Long names are usually used to discourage users from calling the functions, or force them to think carefully.

**Review by yungyuc — APPROVED** [review]
LGTM.  Good work.
